### PR TITLE
Fixes to issues with `bsoProcessPendingChanges`

### PR DESCRIPTION
**Pull request by td202**
## Purpose

This fixes two issues with `bsoProcessPendingChanges`:
- Issue #279: The Persistent state implementation of `redelegatePassive` mistakenly set the delegation to `None`. Another mistake in `bsoProcessPendingChanges` meant that delegators to removed bakers were not correctly moved to the passive pool in the index. (These two issues masked each other, as if only one was present we would expect to see invariant violations.) These are fixed with the revision, which is more explicit about state handling to avoid such mistakes.
- The Basic state implementation of `bsoProcessPendingChanges` did not remove the aggregation keys of removed bakers from the index. That is corrected.

## Changes

The code for `bsoProcessPendingChanges` in both basic and persistent is updated to address the issues in each. The persistent implementation is made more explicit with respect to state updates to ensure that the intention of changes is more obvious.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 1]
```diff
@@ -2391,68 +2426,67 @@ doProcessPendingChanges persistentBS isEffective = do
...
+        _1 .= accounts2
+
+        -- Remove the baker's aggregation key from the aggregation keys set
+        abi <- refLoad (acctBkr ^. accountBakerInfo)
+        (_2 .=) =<< Trie.delete (abi ^. BaseAccounts.bakerAggregationVerifyKey) =<< use _2
+
+        -- Add the delegators to the passive delegators
+        oldPAD <- use _3
+        newDset <- foldM (\t d -> Trie.insert d () t) (adDelegators oldPAD) dlist
+        _3 .= PersistentActiveDelegatorsV1 newDset (adDelegatorTotalCapital oldPAD + dcapital)
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs:2476]: Likewise the above 3 uses of `.=`.

File: concordium-consensus/src/Concordium/GlobalState/Basic/BlockState.hs [hunk 2]
```diff
@@ -1057,29 +1057,31 @@ instance (IsProtocolVersion pv, Monad m) => BS.BlockStateOperations (PureBlockSt
                 preuse (blockAccounts . Accounts.indexedAccount accId) >>= \case
                     Just acct -> case acct ^? accountBaker of
                         Just acctBkr@AccountBaker{..} -> case _bakerPendingChange of
-                            RemoveStake pet | isEffective pet ->
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/GlobalState/Basic/BlockState.hs:1060]: I've refactored `processBakers` so it returns `()` instead, which I think is a cleaner abstraction. Also added some commentary.

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 3]
```diff
@@ -2282,106 +2286,137 @@ doProcessPendingChanges
...
-            bsp{bspBirkParameters = (bspBirkParameters bsp){_birkActiveBakers = newAB}}
-        return total
-
+      processPendingChanges bsp0 = do
+        pab0 <- refLoad (bsp0 ^. birkParameters . birkActiveBakers)
+        (ps1, a1) <- modifyPassiveDelegation (pab0, bspAccounts bsp0)
+        ((pab2, accts2), a2) <- modifyBakers ps1
+        -- Update the total active capital to the sum of the new passive delegators and bakers.
+        newAB <- refMake (pab2 & totalActiveCapital .~ TotalActiveCapitalV1 (a1 + a2)) 
+        return $ bsp0 {bspAccounts = accts2} & (birkParameters . birkActiveBakers .~ newAB)
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Added `$!`

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs [hunk 4]
```diff
@@ -2282,106 +2286,137 @@ doProcessPendingChanges
...
-          -> MTL.WriterT Amount (MTL.StateT (BlockStatePointers pv) m) (Trie.Alteration (PersistentActiveDelegators 'AccountV1))
+          -> MTL.WriterT
+                Amount
+                (MTL.StateT (Accounts.Accounts pv, AggregationKeySet, PersistentActiveDelegators 'AccountV1) m)
+                (Trie.Alteration (PersistentActiveDelegators 'AccountV1))
       processBaker bid@(BakerId accId) oldDelegators = do
-          newDelegators <- lift $ processDelegators oldDelegators
+          accts0 <- use _1
+          (newDelegators, accts1) <- lift $ lift $ MTL.runStateT (processDelegators oldDelegators) accts0
+          _1 .= accts1
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: Unclear, so I've switched to `.=!`

File: concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs (hunk 1 above)
Comment (td202) [reply concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState.hs]: As above.

**Review by eb-concordium — APPROVED** [review]

**Pull request merged**